target/linux/ath79/dts/ar7242_meraki_mr12.dts

```diff
@@ -0,0 +1,180 @@
+#include "ar7242.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "meraki,mr12", "qca,ar7242";
+	model = "Meraki MR12";
+
+	aliases {
+		led-boot = &led_power_orange;
+		led-failsafe = &led_power_orange;
+		led-running = &led_power_green;
+		led-upgrade = &led_power_orange;
+	};
+
+	extosc: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-output-names = "ref";
+		clock-frequency = <40000000>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi1 {
+			label = "green:wifi1";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi2 {
+			label = "green:wifi2";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi3 {
+			label = "green:wifi3";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi4 {
+			label = "green:wifi4";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		wan {
+			label = "green:wan";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_orange: power_orange {
+			label = "orange:power";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			panic-indicator;
+		};
+
+		led_power_green: power_green {
+			label = "green:power";
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&pcie {
+	status = "okay";
+
+	ath9k: wifi@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&config 0x66>;
+		mtd-mac-address-increment = <1>;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+};
+
+&pll {
+	clocks = <&extosc>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy4: ethernet-phy@4 {
+		reg = <4>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+	mtd-mac-address = <&config 0x66>;
+
+        pll-data = <0x16000000 0x00000101 0x00001313>;
+        pll-reg = <0x4 0x2c 17>;
+        pll-handle = <&pll>;
```

Comment:
you only need pll-data, the rest is in the dtsi

if you have tested speeds 100 and 10 already, try this
`        pll-data = <0x02000000 0x00000101 0x00001313>;`
or if that doesn't work, this
`        pll-data = <0x03000000 0x00000101 0x00001313>;`

also watch indentation

Reply:
I'm testing connected to an MX60W with an internal AR8327. Do I want to use ethtool or swconfig to poke the port to the right speed?

Reply:
ethtool would work on the MR12 for eth0 but not eth1 because it is internal, don't worry about speeds on eth1

this is how it is done for a raspberry pi, syntax for openwrt may be different
`ethtool -s eth0 speed 100 duplex full autoneg off`

pinging with large packets is a good way to tell the link is good and which speed you're on depending on the time.

`ping -s 9999 192.168.1.xxx`

Reply:
Speed 10 on the AR8021 phy doesn't work so far; 100 and 1000 are fine. Let's try with your pll data's.

Reply:
in that case you probably want this
`        pll-data = <0x02000000 0x00000101 0x00001616>;`

Reply:
`pll-data = <0x02000000 0x00000101 0x00001313>;` works on 10, 100, and 1000.

Reply:
I thought you just said `0x00001313` for speed 10 didn't work lol

Reply:
I did, it didn't. Could have been one of the other changes I made in that diff too, or the exact invocation of `ethtool`. Let me go back to `pll-data = <0x16000000 0x00000101 0x00001313>;` and see if I can reproduce.

Reply:
FYI theres no difference for speed 10, we have only changed speed 1000, the registers are used as follows

`pll-data = <speed_1000 speed_100 speed_10>;`

Reply:
Yeah, I was thinking either I invoked ethtool differently or there was some other change that misled me -- I was not pinging from the router, but rather `iperf3'ing` through laptop -> eth1 -> eth0 -> mx60.

Must have commented out `pll-data = <0x16000000 0x00000101 0x00001313>;` when testing. `00001616` from the .dtsi does not work on 10Mbps, while `00001313` does. I'll go back to using what the `avm_fritz300e` DTS uses, `pll-data = <0x16000000 0x00000101 0x00001313>;`.

Reply:
> `pll-data = <0x02000000 0x00000101 0x00001313>;` works on 10, 100, and 1000.

stick with this, like you said it is good

this one`<0x16000000` sets extra delay that is now handled by the switch chip and not needed

Reply:
Tested; 1000 works fine with `0x16000000`. Pushed as `f9e91f93a1`.

Reply:
do `0x02000000` instead...it was fine the way it was as long as your test was good...

also no need to keep pushing until we fix everything

Reply:
Got it, I just wanted to establish credence that I was actually building and committing these changes and getting these results ...

Reply:
> Got it, I just wanted to establish credence that I was actually building and committing these changes and getting these results ...

True I worry about the same thing sometimes. I always do `make clean` between builds, even if it's not necessary

Reply:
> `make clean`

Could also be why. You'd think that dts changes would just impact the DTS, but I've only been `make`ing.

Reply:
[pll-data is now `<0x02000000> ...](https://github.com/openwrt/openwrt/pull/3634/commits/e783ef483365f08712c29411b2a73891e5cc2ae8#diff-ebe7e10623c0c66c5d4660984125f80f6ddb6439e5b52f53dc35dd70f48c14ceR109). 10/100/1000 tested. Marking as resolved.